chmigrate/migrator.go

```diff
@@ -149,16 +158,23 @@ func (m *Migrator) Migrate(ctx context.Context, opts ...MigrationOption) (*Migra
 		migration := &group.Migrations[i]
 		migration.GroupID = group.ID
 
-		// Always mark migration as applied so the rollback has a chance to fix the database.
-		if err := m.MarkApplied(ctx, migration); err != nil {
-			return nil, err
-		}
-
-		if !cfg.nop && migration.Up != nil {
-			if err := migration.Up(ctx, m.db); err != nil {
+		var err error
+		if !cfg.nop || migration.Up != nil {
+			err = migration.Up(ctx, m.db)
+			// If the migration failed and we only apply on success, return error
+			if err != nil && m.markAppliedOnSuccess {
 				return group, err
 			}
 		}
+
+		if applyErr := m.MarkApplied(ctx, migration); applyErr != nil {
+			return group, applyErr
+		}
+
+		// If the migration failed, return error
+		if err != nil {
+			return group, err
+		}
```

Comment:
Let's make this simpler and just have `MarkApplied` with corresponding ifs before and after `migration.Up`. It is easier to read and also marks migrations as applied even migrations panic / hang indefinitely.

Reply:
If a migration is marked as applied before the migration is attempted it's possible for a situation where the program exits before the up/down happens and on the next run the migration would be skipped. The more I think about it, the apply on failure path is very undesirable and shouldn't exist, it's just too easy for a developer to put the DB in a bad state and not realize it.

Reply:
It is useful at least for me for the reasons described in the comment.

>It's just too easy for a developer to put the DB in a bad state and not realize it.

The db is in a bad state no matter when we mark a migration as applied. All that is changed is that with the current approach you rollback the migration hoping to fix the damage.

Reply:
Agreed, the DB is in a bad state and requires manual intervention. No way around that. We are down into the nitty-gritty details of an exceptional case.

By failing the migration and not allowing the tool to apply additional migrations we stop ourselves from making the problem worse. Ideally, we execute and do things perfectly, but we are human and our tools should factor in. We simply don't make the perfect decisions or execute processes perfectly when under stress and a migration failing in production is stressful. How often do we press up and enter our terminals instinctively or rerun deploys in case it was a transient problem?

The major problem is running a 2nd up (or down) right after the first one fails, which will run additional migrations and make the situation worse. Once we are in this bad state, IMO, the tool should stop applying migrations (up or down) until the situation is resolved. Even applying downs after a failed up is a problem because the down assumes that the up was entirely successful and if data updates are in the migration it may apply those data changes asymmetrically.

Reply:
I will make the requested changes tomorrow. I have reservations about the value of the mark applied on failure and the "foot-gun" it creates.

Reply:
Changes made.